sparse/src/KokkosSparse_BsrMatrix.hpp

```diff
@@ -912,6 +912,94 @@ class BsrMatrix {
     return const_block_type(&values(i * blockDim_ * blockDim_), blockDim_, blockDim_);
   }
 
+  /// \brief Convert the Bsr into a CrsMatrix
+  ///
+  /// The default return type will be a CrsMatrix with all the same template arguments
+  /// as this Bsr, but you can provide your own type if needed. The only requirement
+  /// is that the execution spaces match.
+  ///
+  /// This is a host function.
+  ///
+  template <typename CrsMatrixType = KokkosSparse::CrsMatrix<ScalarType, OrdinalType, Device, MemoryTraits, SizeType>>
+  CrsMatrixType convertToCrs() const {
+    using crs_size_t      = typename CrsMatrixType::size_type;
+    using crs_rowmap_t    = typename CrsMatrixType::row_map_type::non_const_type;
+    using crs_entries_t   = typename CrsMatrixType::index_type::non_const_type;
+    using crs_values_t    = typename CrsMatrixType::values_type::non_const_type;
+    using crs_exe_space_t = typename CrsMatrixType::execution_space;
+    using policy_t        = typename Kokkos::TeamPolicy<execution_space>;
+    using member_t        = typename policy_t::member_type;
+
+    static_assert(std::is_same_v<crs_exe_space_t, execution_space>,
+                  "We do not currently support converting a BsrMatrix to a CsrMatrix with a different execution space");
+
+    // Get size/dimension info from this Bsr. We will use crs_size_t for all int types
+    const crs_size_t blockDim        = this->blockDim();
+    const crs_size_t blockSize       = blockDim * blockDim;
+    const crs_size_t numBlockRows    = numRows();
+    const crs_size_t numBlockCols    = numCols();
+    const crs_size_t numBlockEntries = nnz();
+
+    // Get graph and values from this Bsr
+    const auto blockRowMap  = graph.row_map;
+    const auto blockEntries = graph.entries;
+    const auto blockValues  = values;
+
+    // Compute Csr row/col/entry sizes by multiplying Bsr sizes by block dimension
+    const crs_size_t numCrsRows    = numBlockRows * blockDim;
+    const crs_size_t numCrsCols    = numBlockCols * blockDim;
+    const crs_size_t numCrsEntries = numBlockEntries * blockSize;
+
+    // Allocate CrsMatrix views
+    crs_rowmap_t crsRowMap("crsRowMap", numCrsRows + 1);
+    crs_entries_t crsEntries("crsEntries", numCrsEntries);
+    crs_values_t crsValues("crsValues", numCrsEntries);
+
+    // Create the policy, we have 3 levels of parallelism available in the algorithm
+    policy_t policy(numBlockRows, Kokkos::AUTO(), blockDim);
```

Comment:
The vector length should be limited to ``policy_t::vector_length_max()``

Reply:
Yep

Reply:
@brian-kelley , done